src/coap_openssl.c

```diff
@@ -492,138 +510,724 @@ int coap_dtls_context_set_psk(coap_context_t *ctx,
+                     ERR_lib_error_string(e),
+                     ERR_func_error_string(e));
+          }
+        }
+      }
+      X509_free(x509);
     }
+    break;
+  default:
+    coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: Unknown key type %d\n",
+             setup_data->key_type);
+    return 0;
   }
-  if (context->tls.ctx) {
-    if (setup_data->public_cert && setup_data->public_cert[0]) {
-      if (!(SSL_CTX_use_certificate_file(context->tls.ctx, setup_data->public_cert, SSL_FILETYPE_PEM))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: TLS: %s: Unable to configure Server Certificate\n", setup_data->public_cert);
+
+  return 1;
+}
+
+static int
+setup_pki_client(SSL *ssl,
+                 coap_dtls_pki_t* setup_data
+) {
+  switch (setup_data->key_type) {
+  case COAP_PKI_KEY_PEM:
+    if (setup_data->key.pem.public_cert && setup_data->key.pem.public_cert[0]) {
+      if (!(SSL_use_certificate_file(ssl, setup_data->key.pem.public_cert,
+            SSL_FILETYPE_PEM))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Client Certificate\n",
+                 setup_data->key.pem.public_cert);
         return 0;
       }
     }
-    else if (setup_data->asn1_public_cert && setup_data->asn1_public_cert_len > 0) {
-      if (!(SSL_CTX_use_certificate_ASN1(context->tls.ctx, setup_data->asn1_public_cert_len, setup_data->asn1_public_cert))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: TLS: %s: Unable to configure Server Certificate\n", "ASN1");
+    else {
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Client Certificate defined\n");
+      return 0;
+    }
+    if (setup_data->key.pem.private_key && setup_data->key.pem.private_key[0]) {
+      if (!(SSL_use_PrivateKey_file(ssl, setup_data->key.pem.private_key,
+            SSL_FILETYPE_PEM))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Client Private Key\n",
+                  setup_data->key.pem.private_key);
         return 0;
       }
     }
     else {
-      coap_log(LOG_ERR, "*** coap_dtls_context_set_pki: TLS: No Server Certificate defined\n");
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Client Private Key defined\n");
+      return 0;
+    }
+    if (setup_data->key.pem.ca_file && setup_data->key.pem.ca_file[0]) {
+      X509_STORE *st;
+      long e;
+      BIO *in;
+      X509 *x = NULL;
+      char *rw_var = NULL;
+      SSL_CTX *ctx = SSL_get_SSL_CTX(ssl);
+
+      st = SSL_CTX_get_cert_store(ctx);
+      in = BIO_new(BIO_s_file());
+      /* Need to do this to not get a compiler warning about const parameters */
+      memcpy(&rw_var, &setup_data->key.pem.ca_file, sizeof (rw_var));
+      if (!BIO_read_filename(in, rw_var)) {
+        BIO_free(in);
+        X509_free(x);
+        break;
+      }
+
+      /* Flush out existing errors */
+      while ((e = ERR_get_error()) != 0) { }
```

Comment:
For better readability, please move the closing parenthesis to the next line.

Reply:
Fixed